In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm

data=pd.read_csv('2.01. Admittance.csv')
data

SAT Admitted
0    1363       No
1    1792      Yes
2    1954      Yes
3    1653       No
4    1593       No
..    ...      ...
163  1722      Yes
164  1750      Yes
165  1555       No
166  1524       No
167  1461       No

[168 rows x 2 columns]

Tenemos 168 filas en este data set, que nos revela aquellos registros que fueron admitidos o no a la universidad segun su promedio SAT

In [2]:
df=data.copy()
df['Admitted']=df['Admitted'].map({'Yes':1, 'No':0})

In [3]:
df

SAT  Admitted
0    1363         0
1    1792         1
2    1954         1
3    1653         0
4    1593         0
..    ...       ...
163  1722         1
164  1750         1
165  1555         0
166  1524         0
167  1461         0

[168 rows x 2 columns]

Variables

- SAT: Variable independiente
- Admitted: Variable dependiente


In [4]:
y= df['Admitted']
X= df['SAT']

In [5]:
x=sm.add_constant(X)
regresion=sm.Logit(y,x).fit()

Optimization terminated successfully.
         Current function value: 0.137766
         Iterations 10


Tomo 10 iteraciones ajustar el modelo

Nota: El numero maximo de iteraciones que hará statsmodels es 35, si suceden más abrá un mensaje de advertencia

In [6]:
regresion.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:               Admitted   No. Observations:                  168
Model:                          Logit   Df Residuals:                      166
Method:                           MLE   Df Model:                            1
Date:                Mon, 10 Jul 2023   Pseudo R-squ.:                  0.7992
Time:                        21:21:49   Log-Likelihood:                -23.145
converged:                       True   LL-Null:                       -115.26
Covariance Type:            nonrobust   LLR p-value:                 5.805e-42
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const        -69.9128     15.737     -4.443      0.000    -100.756     -39.070
SAT            0.0420      0.009      4.454      0.000       0.024       0.060
==============================================================================

Possibly complete quasi-separation: A fraction 0.27 of observations can be
perfectly predicted. This might indicate that there is complete
quasi-separation. In this case some parameters will not be identified.
"""


Revisemos los resultados...


MLE o EMV es la estimación maxima de verosimilitud

Para aclarar, la función de verosimilitud estima la posibilidad de que un modelo describa la verdadera relación subyacente entre las variables, entonces a mayor verosimilitud. En este caso la MLE intenta maximizar la verosimilitud, hasta el punto en que ya no puede optimizarlo más, entonces ahí se detiene.

Log-Likelihood

Este valor indica la verosimilitud y casi siempre es negativo aunque no siempre. Mientras mayor sea, mejor será

LL-Null

Este es el valor de la verosimilitud pero sin variables independientes

LLR p-value

Lo que hace este valor es comprar el Log-likelihood con la LL-Null, permitiendonos conocer si nuestro modelo es inutil. El valor LLR p indica la significancia estadística de la diferencia entre los dos modelos comparados. Un valor p bajo (por debajo de un umbral predefinido, como 0.05) sugiere que el modelo completo proporciona una mejora significativa en la bondad de ajuste en comparación con el modelo restringido, lo que respalda la inclusión de las variables o restricciones adicionales en el modelo completo.

Pseudo R squared

Las medidas de pseudo R-cuadrado tienen como objetivo proporcionar una estimación de la proporción de varianza explicada por el modelo en el contexto de resultados binarios o categóricos. Sin embargo, es importante tener en cuenta que existen diferentes tipos de medidas de pseudo R-cuadrado, y cada una tiene su propio método de cálculo e interpretación. R-cuadrado de McFadden (prsquared): El R-cuadrado de McFadden es una de las medidas de pseudo R-cuadrado más utilizadas para la regresión logística. Se define como la diferencia entre el logaritmo de verosimilitud del modelo completo y el modelo nulo dividido por el logaritmo de verosimilitud del modelo nulo. El valor resultante varía de 0 a 1, donde los valores más altos indican un mejor ajuste. Así que un adecuado R cuadrado estaría entre 0.4 y 0.2

La ecuación que esta funcionando aquí para predecir la posibilidad de que ocurra un evento sería

*log(π/1-π)= -69,9128+0.0420*SAT*

Siendo

*π = probabilidad de que ocurra el evento*

*1-π = probabilidad de que un evento no ocurre*